File: cli/scripts/split_point.py
#!/usr/bin/python
#
# Originally written by Davy Temperley in Perl.
#

import math
from pathlib import Path
import sys
from typing import List, Optional, Tuple


try:
    file = Path(sys.argv[1])
except IndexError:
    print('Filename should be only arg')
    sys.exit(1)

if not file.is_file():
    raise FileNotFoundError

vals = []
with file.open() as f:
    for line in f:
        val = line.strip('\n')
        try:
            vals.append(float(val))
        except ValueError:
            print(f'{val} is not a float')
            sys.exit(1)


CUSHION = 1

best_dev_sum = float('inf')
best_dev_split_point: Optional[int] = None
best_mean1: Optional[int] = None
best_mean2: Optional[int] = None


def sum_sq_dev(li: List[int]) -> Tuple[int, int]:
    avg = sum(li) / len(li)
    dev = 0
    for v in li:
        dev += (v - avg) ** 2
    return dev, avg


for i in range(CUSHION, len(vals) - CUSHION):
    first_half = vals[:i + 1]
    second_half = vals[i + 1:]

    dev1, avg1 = sum_sq_dev(first_half)
    dev2, avg2 = sum_sq_dev(second_half)
    dev_sum = dev1 + dev2

    if dev_sum < best_dev_sum:
        best_dev_sum = dev_sum
        best_dev_split_point = i
        best_mean1 = avg1
        best_mean2 = avg2


print(f'''
    Min summed squared deviations = {best_dev_sum:.2f}
    Split point = {best_dev_split_point}
    Mean1 = {best_mean1:.2f}
    Mean2 = {best_mean2:.2f}
''')

File: cli/scripts/split-point.pl
#!/usr/bin/perl

while(<>) {

    chomp($_);
    @y[$i] = $_;
    # printf("%.3f\n", $y[$i]);
    $i++;

}

$cushion = 1;
$max_y = $i-1;
$min_y = 0;
$max_s = $max_y - $cushion;
$min_s = $cushion;

$best_difference = $best_abs_log_ratio = 0;
$best_dev_sum = 1000000;

# printf("max_y = %d\n", $max_y);

for($s=$min_s; $s<=$max_s; $s++) {

    $mass1 = $mass2 = $dev1 = $dev2 = 0;

    for($i2=$min_y; $i2<=$s; $i2++) {      # Include split point in the first half
	$mass1 += $y[$i2];
    }
    $avg1 = $mass1/(($s-$min_y)+1);
    for($i2=$min_y; $i2<=$s; $i2++) {      # Include split point in the first half
	$dev1 += ($y[$i2] - $avg1) ** 2;
    }

    for($i2=$s+1; $i2<=$max_y; $i2++) {
	$mass2 += $y[$i2];
    }
    $avg2 = $mass2/($max_y - $s);
    for($i2=$s+1; $i2<=$max_y; $i2++) {
	$dev2 += ($y[$i2] - $avg2) ** 2;
    }
    
    $difference = $avg2 - $avg1;
    $log_ratio = log($avg2 / $avg1);
    $dev_sum = $dev1 + $dev2;
    printf("%d: avg1 = %.3f; avg2 = %.3f; diff = %.3f, ratio = %.3f, log ratio = %.3f, dev1 = %.3f, dev2 = %.3f, dev sum = %.3f\n", $s, $avg1, $avg2, $difference, $avg2 / $avg1, $log_ratio, $dev1, $dev2, $dev_sum);
    if(abs($difference) > $best_abs_difference) {
	$best_abs_difference = abs($difference);
	$best_difference = $difference;
	$best_diff_split_point = $s;
    }
    if(abs($log_ratio) > $best_abs_log_ratio) {
	$best_abs_log_ratio = abs($log_ratio);
	$best_log_ratio = $log_ratio;
	$best_ratio_split_point = $s;
    }
    if($dev_sum < $best_dev_sum) {
	$best_dev_sum = $dev_sum;
	$best_dev_split_point = $s;
	$best_mean1 = $avg1;
	$best_mean2 = $avg2;
    }
}

printf("Max absolute difference = %.3f (%.3f), split point = %d\n", $best_abs_difference, $best_difference, $best_diff_split_point);
printf("Max abs log ratio = %.3f (%.3f), split point = %d\n", $best_abs_log_ratio, $best_log_ratio, $best_ratio_split_point);
printf("Min summed squared deviations = %.3f, split point = %d, mean1 = %.3f, mean2 = %.3f\n", $best_dev_sum, $best_dev_split_point, $best_mean1, $best_mean2);
